File: python/nested_functions.py
"""
Python test file for nested functions.

This file contains examples of nested function definitions
and closures to test DiffScope's ability to detect changes
in nested function structures.
"""

def outer_function_simple(x):
    """
    A simple outer function with a nested function.
    
    Args:
        x: Input value
        
    Returns:
        Result of inner function
    """
    def inner_function(y):
        """Inner function that adds x and y."""
        return x + y
    
    return inner_function(10)


def create_counter():
    """
    Create a counter with increment, decrement, and get methods.
    
    Returns:
        A dictionary containing counter methods
    """
    count = 0
    
    def increment():
        """Increment counter by 1."""
        nonlocal count
        count += 1
        return count
    
    def decrement():
        """Decrement counter by 1."""
        nonlocal count
        count -= 1
        return count
    
    def get_count():
        """Get current counter value."""
        return count
    
    def reset():
        """Reset counter to 0."""
        nonlocal count
        count = 0
        return count
    
    return {
        "increment": increment,
        "decrement": decrement,
        "get_count": get_count,
        "reset": reset
    }


def create_adder(x):
    """
    Create a function that adds x to its argument.
    
    Args:
        x: Value to add
        
    Returns:
        A function that adds x to its argument
    """
    def adder(y):
        """Add x to y."""
        return x + y
    
    return adder


def create_multiplier(x):
    """
    Create a function that multiplies its argument by x.
    
    Args:
        x: Multiplication factor
        
    Returns:
        A function that multiplies its argument by x
    """
    def multiplier(y):
        """Multiply y by x."""
        return x * y
    
    return multiplier


def create_power_function(exponent):
    """
    Create a function that raises its argument to the given exponent.
    
    Args:
        exponent: The exponent to use
        
    Returns:
        A function that raises its argument to the given exponent
    """
    def power_function(base):
        """Raise base to exponent."""
        return base ** exponent
    
    return power_function


def create_sequence_processor():
    """
    Create a collection of sequence processing functions.
    
    Returns:
        A dictionary containing sequence processing functions
    """
    def map_sequence(sequence, func):
        """
        Apply func to each element of the sequence.
        
        Args:
            sequence: Input sequence
            func: Function to apply
            
        Returns:
            List containing func applied to each element
        """
        return [func(item) for item in sequence]
    
    def filter_sequence(sequence, predicate):
        """
        Filter sequence by predicate.
        
        Args:
            sequence: Input sequence
            predicate: Function that returns True/False
            
        Returns:
            List containing elements for which predicate is True
        """
        return [item for item in sequence if predicate(item)]
    
    def reduce_sequence(sequence, func, initial=None):
        """
        Reduce sequence using func and initial value.
        
        Args:
            sequence: Input sequence
            func: Binary function to apply cumulatively
            initial: Initial value (default: None)
            
        Returns:
            Accumulated result
        """
        result = initial
        for item in sequence:
            if result is None:
                result = item
            else:
                result = func(result, item)
        return result
    
    return {
        "map": map_sequence,
        "filter": filter_sequence,
        "reduce": reduce_sequence
    }


def function_with_multiple_levels():
    """
    A function with multiple levels of nested functions.
    
    Returns:
        A function that creates nested functions
    """
    def level1(x):
        """First level nested function."""
        def level2(y):
            """Second level nested function."""
            def level3(z):
                """Third level nested function."""
                return x + y + z
            return level3
        return level2
    return level1


if __name__ == "__main__":
    # Test simple nested function
    assert outer_function_simple(5) == 15
    
    # Test counter
    counter = create_counter()
    assert counter["get_count"]() == 0
    counter["increment"]()
    counter["increment"]()
    assert counter["get_count"]() == 2
    counter["decrement"]()
    assert counter["get_count"]() == 1
    counter["reset"]()
    assert counter["get_count"]() == 0
    
    # Test adder and multiplier
    add_five = create_adder(5)
    assert add_five(10) == 15
    
    double = create_multiplier(2)
    assert double(7) == 14
    
    # Test power function
    square = create_power_function(2)
    cube = create_power_function(3)
    assert square(4) == 16
    assert cube(2) == 8
    
    # Test sequence processor
    seq_proc = create_sequence_processor()
    numbers = [1, 2, 3, 4, 5]
    assert seq_proc["map"](numbers, lambda x: x * 2) == [2, 4, 6, 8, 10]
    assert seq_proc["filter"](numbers, lambda x: x % 2 == 0) == [2, 4]
    assert seq_proc["reduce"](numbers, lambda x, y: x + y) == 15
    
    # Test multiple levels
    level1_func = function_with_multiple_levels()
    level2_func = level1_func(10)
    level3_func = level2_func(20)
    assert level3_func(30) == 60

File: javascript/nested_functions.js
/**
 * JavaScript test file for nested functions.
 * 
 * This file contains examples of nested function definitions
 * and closures to test DiffScope's ability to detect changes
 * in nested function structures.
 */

/**
 * A simple outer function with a nested function.
 * @param {number} x - Input value
 * @returns {number} Result of inner function
 */
function outerFunctionSimple(x) {
  /**
   * Inner function that adds x and y.
   * @param {number} y - Value to add to x
   * @returns {number} Sum of x and y
   */
  function innerFunction(y) {
    return x + y;
  }
  
  return innerFunction(10);
}

/**
 * Create a counter with increment, decrement, and get methods.
 * @returns {Object} An object containing counter methods
 */
function createCounter() {
  let count = 0;
  
  /**
   * Increment counter by 1.
   * @returns {number} New count value
   */
  function increment() {
    return ++count;
  }
  
  /**
   * Decrement counter by 1.
   * @returns {number} New count value
   */
  function decrement() {
    return --count;
  }
  
  /**
   * Get current counter value.
   * @returns {number} Current count value
   */
  function getCount() {
    return count;
  }
  
  /**
   * Reset counter to 0.
   * @returns {number} Reset count value (0)
   */
  function reset() {
    count = 0;
    return count;
  }
  
  return {
    increment,
    decrement,
    getCount,
    reset
  };
}

/**
 * Create a function that adds x to its argument.
 * @param {number} x - Value to add
 * @returns {Function} A function that adds x to its argument
 */
function createAdder(x) {
  /**
   * Add x to y.
   * @param {number} y - Value to add to x
   * @returns {number} Sum of x and y
   */
  return function adder(y) {
    return x + y;
  };
}

/**
 * Create a function that multiplies its argument by x.
 * @param {number} x - Multiplication factor
 * @returns {Function} A function that multiplies its argument by x
 */
function createMultiplier(x) {
  /**
   * Multiply y by x.
   * @param {number} y - Value to multiply by x
   * @returns {number} Product of x and y
   */
  return function multiplier(y) {
    return x * y;
  };
}

/**
 * Create a function that raises its argument to the given exponent.
 * @param {number} exponent - The exponent to use
 * @returns {Function} A function that raises its argument to the given exponent
 */
function createPowerFunction(exponent) {
  /**
   * Raise base to exponent.
   * @param {number} base - Value to raise to the exponent
   * @returns {number} base raised to exponent
   */
  return function powerFunction(base) {
    return Math.pow(base, exponent);
  };
}

/**
 * Create a collection of sequence processing functions.
 * @returns {Object} An object containing sequence processing functions
 */
function createSequenceProcessor() {
  /**
   * Apply func to each element of the sequence.
   * @param {Array} sequence - Input sequence
   * @param {Function} func - Function to apply
   * @returns {Array} Array containing func applied to each element
   */
  function mapSequence(sequence, func) {
    return sequence.map(func);
  }
  
  /**
   * Filter sequence by predicate.
   * @param {Array} sequence - Input sequence
   * @param {Function} predicate - Function that returns true/false
   * @returns {Array} Array containing elements for which predicate is true
   */
  function filterSequence(sequence, predicate) {
    return sequence.filter(predicate);
  }
  
  /**
   * Reduce sequence using func and initial value.
   * @param {Array} sequence - Input sequence
   * @param {Function} func - Binary function to apply cumulatively
   * @param {*} [initial] - Initial value (optional)
   * @returns {*} Accumulated result
   */
  function reduceSequence(sequence, func, initial) {
    return initial !== undefined
      ? sequence.reduce(func, initial)
      : sequence.reduce(func);
  }
  
  return {
    map: mapSequence,
    filter: filterSequence,
    reduce: reduceSequence
  };
}

/**
 * A function with multiple levels of nested functions.
 * @returns {Function} A function that creates nested functions
 */
function functionWithMultipleLevels() {
  /**
   * First level nested function.
   * @param {number} x - First value
   * @returns {Function} Second level function
   */
  return function level1(x) {
    /**
     * Second level nested function.
     * @param {number} y - Second value
     * @returns {Function} Third level function
     */
    return function level2(y) {
      /**
       * Third level nested function.
       * @param {number} z - Third value
       * @returns {number} Sum of x, y, and z
       */
      return function level3(z) {
        return x + y + z;
      };
    };
  };
}

// Module pattern (IIFE with closures)
const calculatorModule = (function() {
  // Private variables
  let result = 0;
  const history = [];
  
  // Private function
  function saveToHistory(operation, value) {
    history.push({ operation, value, result });
  }
  
  // Public API
  return {
    add: function(x) {
      result += x;
      saveToHistory('add', x);
      return result;
    },
    subtract: function(x) {
      result -= x;
      saveToHistory('subtract', x);
      return result;
    },
    multiply: function(x) {
      result *= x;
      saveToHistory('multiply', x);
      return result;
    },
    divide: function(x) {
      if (x === 0) throw new Error("Cannot divide by zero");
      result /= x;
      saveToHistory('divide', x);
      return result;
    },
    getResult: function() {
      return result;
    },
    reset: function() {
      result = 0;
      return result;
    },
    getHistory: function() {
      return [...history]; // Return a copy
    }
  };
})();

// Test cases
if (require.main === module) {
  // Test simple nested function
  console.assert(outerFunctionSimple(5) === 15, "outerFunctionSimple test");
  
  // Test counter
  const counter = createCounter();
  console.assert(counter.getCount() === 0, "counter initial value");
  counter.increment();
  counter.increment();
  console.assert(counter.getCount() === 2, "counter after increment");
  counter.decrement();
  console.assert(counter.getCount() === 1, "counter after decrement");
  counter.reset();
  console.assert(counter.getCount() === 0, "counter after reset");
  
  // Test adder and multiplier
  const addFive = createAdder(5);
  console.assert(addFive(10) === 15, "addFive test");
  
  const double = createMultiplier(2);
  console.assert(double(7) === 14, "double test");
  
  // Test power function
  const square = createPowerFunction(2);
  const cube = createPowerFunction(3);
  console.assert(square(4) === 16, "square test");
  console.assert(cube(2) === 8, "cube test");
  
  // Test sequence processor
  const seqProc = createSequenceProcessor();
  const numbers = [1, 2, 3, 4, 5];
  console.assert(
    JSON.stringify(seqProc.map(numbers, x => x * 2)) === JSON.stringify([2, 4, 6, 8, 10]),
    "map test"
  );
  console.assert(
    JSON.stringify(seqProc.filter(numbers, x => x % 2 === 0)) === JSON.stringify([2, 4]),
    "filter test"
  );
  console.assert(seqProc.reduce(numbers, (x, y) => x + y) === 15, "reduce test");
  
  // Test multiple levels
  const level1Func = functionWithMultipleLevels();
  const level2Func = level1Func(10);
  const level3Func = level2Func(20);
  console.assert(level3Func(30) === 60, "multiple levels test");
  
  // Test calculator module
  console.assert(calculatorModule.getResult() === 0, "calculator initial value");
  calculatorModule.add(5);
  calculatorModule.multiply(2);
  console.assert(calculatorModule.getResult() === 10, "calculator after operations");
  
  console.log("All tests passed!");
}

module.exports = {
  outerFunctionSimple,
  createCounter,
  createAdder,
  createMultiplier,
  createPowerFunction,
  createSequenceProcessor,
  functionWithMultipleLevels,
  calculatorModule
};
